terminal/bundles/org.eclipse.terminal.view.ui/src/org/eclipse/terminal/view/ui/internal/TerminalService.java

```diff
@@ -195,8 +195,18 @@ private CompletableFuture<?> executeServiceOperation(final TerminalServiceRunnab
 			final String title, final ITerminalConnector connector, final Object data) {
 		// Execute the operation
 		if (runnable.isExecuteAsync()) {
-			return CompletableFuture.supplyAsync(() -> runnable.run(tvid, title, connector, data),
-					Display.getDefault());
+			try {
+				return CompletableFuture.supplyAsync(() -> runnable.run(tvid, title, connector, data),
+						/*
+						 * A special executor is used here because the default Display executor will
+						 * do a syncExec here, but isExecuteAsync is true, so we need to return from
+						 * executeServiceOperation *before* we execute the runnable.
+						 */
+						Display.getDefault()::asyncExec);
+			} catch (RuntimeException e) {
+				return CompletableFuture.failedFuture(e);
+			}
```

Comment:
This must be moved a bit further so it covers the whole method.

Assume Line 211 throws an exception then it will bubble up here, while it is more semantic to be reported through the `CompletableFuture`, so callers will never get a (direct) exception here.

Reply:
Done.

FWIW in practice this code is still overly complicated - nothing ever examines the completedfuture returned and `isExecuteAsync` always returns true in practice (in at least the code base I can see)